### PR TITLE
Add support to realtime probes in HDRP

### PR DESCRIPTION
**Pull request by fredpointzero**
### Purpose of this PR
This avoid recusrive rendering inside the SRP Rneder call.

The spirit of the PR is to move from:
 - For each camera
    1. Render recursively all realtime probes (now forbidden)
    2. Render the camera
To:
  1. Compute parameters and cull all cameras
  2. Build a list of visible realtime probes to render
  3. Build a list of render request to perform, rendering probes first
  4. Render all requests

---
### Release Notes
Added support to realtime probes
Render only visible realtime probes

---
### Testing status
**Katana Tests**: First off we need to make sure the Katana SRP tests are green?

**Manual Tests**: What did you do?

**Automated Tests**: What did you setup?

Any test projects to go with this to help reviewers?

---
### Overall Product Risks
**Technical Risk**: High
 - Basically the render loop was refactored

**Halo Effect**: High
 - Anyithing rendered may have been impacted

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/HDProbeUI.Drawers.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-//#define ENABLE_BAKED_PLANAR
+#define ENABLE_BAKED_PLANAR
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/Reflection/HDProbeUI.Drawers.cs]: it should stay commented right ?

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/Reflection/HDProbeSystem.cs [hunk 2]
```diff
@@ -246,20 +257,39 @@ internal void UnregisterProbe(HDProbe probe)
...
+            Assert.IsNotNull(state.cullingGroup, "Culling was not prepared, please prepare cull before performing it.");
+            Assert.IsNotNull(state.hdProbes, "Culling was not prepared, please prepare cull before performing it.");
+            var stateHash = ComputeStateHashDebug(m_PlanarProbeBounds, m_PlanarProbes, m_PlanarProbeCount);
+            Assert.AreEqual(stateHash, state.stateHash, "HDProbes changes since culling was prepared, this will lead to incorrect results.");
+
+            results.Reset();
+            var probes = results.writeableVisibleProbes;
+
+            // TODO: Pool this array to avoid GC (length must be provided with ReflectionSystemParameters from HDRenderPipeline)
+            var indices = new int[1024];
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/Reflection/HDProbeSystem.cs]: I am not sure to understand what is the problem, we know the max number of realtime reflection, why not using it ?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 3]
```diff
@@ -831,704 +858,1092 @@ protected override void Render(ScriptableRenderContext renderContext, Camera[] c
...
                     {
-                        VolumeManager.instance.Update(hdCamera.volumeAnchor, hdCamera.volumeLayerMask);
+                        // Don't add it if it has already been updated this frame or not a real time probe
+                        // TODO: discard probes that are baked once per frame and already baked this frame
+                        if (probe.mode != ProbeSettings.Mode.Realtime)
+                            return;
+
+                        if (!visibleProbes.TryGetValue(probe, out List<int> visibleInIndices))
+                        {
+                            visibleInIndices = new List<int>();// TODO: Pool this to avoid GC
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs]: gc

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 4]
```diff
@@ -831,704 +858,1092 @@ protected override void Render(ScriptableRenderContext renderContext, Camera[] c
...
-                    m_LightLoop.NewFrame(currentFrameSettings);
+                    for (int j = 0; j < cameraSettings.Count; ++j)
+                    {
+                        Camera camera = m_ProbeCameraPool.Count == 0
+                            ? camera = new GameObject().AddComponent<Camera>()
+                            : camera = m_ProbeCameraPool.Pop();
+
+                        camera.gameObject.hideFlags = HideFlags.HideAndDontSave;
+                        camera.gameObject.SetActive(false);
+                        camera.name = $"HDProbe RenderCamera ({visibleProbe.name}:{j} for {viewerTransform})";
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs]: gc

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 5]
```diff
@@ -831,704 +858,1092 @@ protected override void Render(ScriptableRenderContext renderContext, Camera[] c
...
+                        //   or Texture2DArray)
+                        //   To do so, we need to first allocate in the cache the location of the target and then assign
+                        //   it here.
+                        var request = new RenderRequest
+                        {
+                            hdCamera = hdCamera,
+                            cullingResults = _cullingResults,
+                            postProcessLayer = postProcessLayer,
+                            destroyCamera = true,
+                            dependsOnRenderRequestIndices = new List<int>(), // TODO: Pool this to avoid GC
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs]: gc

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 6]
```diff
@@ -831,704 +858,1092 @@ protected override void Render(ScriptableRenderContext renderContext, Camera[] c
...
+                bool haveAsyncTaskWithShadows = false;
+                if (hdCamera.frameSettings.BuildLightListRunsAsync())
+                {
+                    buildLightListTask.Start(cmd, renderContext, (CommandBuffer asyncCmd) =>
+                    {
+                        m_LightLoop.BuildGPULightListsCommon(hdCamera, asyncCmd, m_SharedRTManager.GetDepthStencilBuffer(hdCamera.frameSettings.enableMSAA), m_SharedRTManager.GetStencilBufferCopy(), m_SkyManager.IsLightingSkyValid());
+                    }, !haveAsyncTaskWithShadows);
 
-                            haveAsyncTaskWithShadows = true;
-
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs]: targetid

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 7]
```diff
@@ -831,704 +858,1092 @@ protected override void Render(ScriptableRenderContext renderContext, Camera[] c
...
-                        // Make sure we are in sync with the debug menu for the msaa count
-                        m_MSAASamples = m_DebugDisplaySettings.data.msaaSamples;
-                        m_SharedRTManager.SetNumMSAASamples(m_MSAASamples);
+                    if (!skipRequest
+                        && additionalCameraData != null
+                        && additionalCameraData.hasCustomRender)
+                    {
+                        skipRequest = true;
+                        // Execute custom render
+                        additionalCameraData.ExecuteCustomRender(renderContext, hdCamera);
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs]: Culling should not be done for cameras with custom render (nothing at all should be done for that matter).
Need to early out before that.

File: com.unity.render-pipelines.high-definition/Editor/RenderPipeline/HDRenderPipelineUI.cs [hunk 8]
```diff
@@ -24,9 +24,9 @@ enum Expandable
         static readonly GUIContent renderPipelineResourcesContent = EditorGUIUtility.TrTextContent("Render Pipeline Resources", "Set of resources that need to be loaded when creating stand alone");
         static readonly GUIContent renderPipelineEditorResourcesContent = EditorGUIUtility.TrTextContent("Render Pipeline Editor Resources", "Set of resources that need to be loaded for working in editor");
         static readonly GUIContent diffusionProfileSettingsContent = EditorGUIUtility.TrTextContent("Diffusion Profile List");
-        //static readonly GUIContent enableShaderVariantStrippingContent = EditorGUIUtility.TrTextContent("Shader Variant Stripping");
-        static readonly GUIContent enableSRPBatcher = EditorGUIUtility.TrTextContent("SRP Batcher (experimental)");
-        static readonly GUIContent enableVariantStrippingLog = EditorGUIUtility.TrTextContent("Shader Variant Log");
+        //static readonly GUIContent enableShaderVariantStrippingContent = EditorGUIUtility.TrTextContent("Enable Shader Variant Stripping");
+        static readonly GUIContent enableSRPBatcher = EditorGUIUtility.TrTextContent("Enable SRP Batcher (experimental)");
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/RenderPipeline/HDRenderPipelineUI.cs]: error here!

File: com.unity.render-pipelines.high-definition/Editor/RenderPipeline/Settings/RenderPipelineSettingsUI.cs [hunk 9]
```diff
@@ -101,9 +101,9 @@ static void Drawer_SectionPrimarySettings(SerializedRenderPipelineSettings d, Ed
             using (new EditorGUI.DisabledScope(!msaaAllowed))
             {
                 ++EditorGUI.indentLevel;
-                EditorGUILayout.PropertyField(d.MSAASampleCount, k_MSAASampleCountContent);
-                --EditorGUI.indentLevel;
-            }
+                    EditorGUILayout.PropertyField(d.MSAASampleCount, k_MSAASampleCountContent);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/RenderPipeline/Settings/RenderPipelineSettingsUI.cs]: weird ident
